akd_quorum/src/crypto.rs

```diff
@@ -0,0 +1,421 @@
+    /// We take ownership of the quorum key here to help prevent leakage of the key.
+    /// By taking ownership, someone needs to explicitely clone it to use it elsewhere
+    fn generate_shards(
+        quorum_key: [u8; QUORUM_KEY_SIZE],
+        num_shards: u8,
+    ) -> Result<Vec<QuorumKeyShard>, QuorumOperationError> {
+        let num_approvals = Self::shards_required(num_shards);
+
+        let mut parts = vec![vec![]; num_shards.into()];
+
+        for i in 0..QUORUM_KEY_NUM_PARTS {
+            let part: [u8; DATA_SIZE] = quorum_key[i * DATA_SIZE..(i + 1) * DATA_SIZE]
+                .try_into()
+                .map_err(|_| {
+                QuorumOperationError::Sharding(format!(
+                    "Unable to convert quorum key slice into SSS shardable component of len {}",
+                    DATA_SIZE
+                ))
+            })?;
+            let results = create_shares(&part, num_shards, num_approvals)?;
+            for node_i in 0..num_shards {
+                let idx: usize = node_i.into();
+                match results.get(idx) {
+                    None => {
+                        return Err(QuorumOperationError::Sharding(format!(
+                            "Resulting shards did not have an shard at entry {}",
+                            node_i
+                        )));
+                    }
+                    Some(part) => {
+                        parts[idx].push(part.clone());
+                    }
+                }
+            }
+        }
+
+        let formatted_shards = QuorumKeyShard::build_from_vec_vec_vec(parts)?;
+        Ok(formatted_shards)
+    }
+
+    /// Reconstruct the quorum key from a specific collection of shards
+    fn reconstruct_shards(
+        shards: Vec<QuorumKeyShard>,
+    ) -> Result<[u8; QUORUM_KEY_SIZE], QuorumOperationError> {
+        let mut potential_result = [0u8; QUORUM_KEY_SIZE];
+        // there should be QUORUM_KEY_NUM_PARTS in each shard
+        for i in 0..QUORUM_KEY_NUM_PARTS {
+            let part_i = shards
+                .iter()
+                .map(|shard| shard.components[i].to_vec())
+                .collect::<Vec<_>>();
+            let some_key = combine_shares(&part_i)?;
+            if let Some(key) = some_key {
+                let deconstructed_partial: [u8; DATA_SIZE] = key.try_into().map_err(|_| QuorumOperationError::Sharding(format!("Reconstructing the quorum key resulted in an invalid key length. It _MUST_ be of length {} bytes", DATA_SIZE)))?;
+                potential_result[i * DATA_SIZE..(i + 1) * DATA_SIZE]
+                    .clone_from_slice(&deconstructed_partial);
+            } else {
+                return Err(QuorumOperationError::Sharding(
+                    "Sharding request to recombine shares resulted in no constructed quorum key"
+                        .to_string(),
+                ));
+            }
+        }
+        Ok(potential_result)
+    }
+}
+
+#[cfg(test)]
+mod crypto_tests {
```

Comment:
I'd propose to move tests into separate files. It is not such a big difference from the code organization point of view, but it is easier to review code if the tests are separated - if the change is only in crypto_tests.rs, than it is clear that the production code is not changed.